HPCC-28687 Add metric naming restrictions

kenrowland opened the pull request:
Added restrictions to metric framework to only accept valid names

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexisnexisrisk.com

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-28687 Add metric naming restrictions

Added restrictions to metric framework to only accept valid names
Moved metric name generation to registration method in ESPCommon

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexisnexisrisk.com

kenrowland commented:
@afishbeck  You will notice a change to httpbinding. This makes sure the process name is actually set. During debug, I found that the constructor would exit prior to storing the name. The name is used in the generated code's binding constructor as a piece of the matric name.

Also, there is a change to some dali code to change the names of the current metrics created to meet the naming restrictions.

afishbeck commented on system/jlib/jmetrics.cpp:
```diff
@@ -138,35 +139,49 @@ bool MetricsManager::addMetric(const std::shared_ptr<IMetric> &pMetric)
-        {
 #ifdef _DEBUG
-            throw MakeStringException(MSGAUD_operator, "addMetric - Attempted to add duplicate named metric with name '%s'", name.c_str());
+        throw MakeStringException(MSGAUD_operator, "addMetric - Attempted to add metric with invalid name ('%s')", name.c_str());
```
Hmmm, I don't usually throw only in debug.  I'm often trying to reproduce behavior found in non-debug builds while stepping through or increasing logging by running debug.  This could get really confusing.  Or if I didn't know how to configure something to get past something like this I would have to modify the code to keep debugging.  We do sometimes have asserts though that only happen in debug.  Might want to see what Gavin thinks.

kenrowland replied:
Only throwing in debug was a design choice made by Gavin and myself. The reason is we wanted the use to know during development that they have provided an invalid metric and to halt system execution. However, in a production environment, we did not want to halt.

afishbeck commented on system/jlib/jmetrics.cpp:
```diff
@@ -138,35 +139,49 @@ bool MetricsManager::addMetric(const std::shared_ptr<IMetric> &pMetric)
-        name.append(".").append(metaDataIt.value);
-    }
+    std::regex pattern("^[A-Za-z][A-Za-z0-9.]*[^.]$");
 
```
Is the last character too permissive?

afishbeck replied:
I think the changed line is good.

afishbeck commented on testing/unittests/metrics/MetricFrameworkTests.cpp:
```diff
@@ -79,10 +80,51 @@ class MetricFrameworkTests : public CppUnit::TestFixture
+            {"9startswithnumber"},
+            {"has_underscore"},
+            {"has&^%$#@!-~`+=|[]{}?/<>,_specialcharacters"}
+        };
```
I think "has.a.&" would currently pass?  One special character at the end of the string?

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -5757,26 +5792,59 @@ void EspServInfo::write_esp_binding()
+        outs("{\n");
+        outs("\tstd::string processName(queryProcessName());\n");  // process/app name
+        outs("\tstd::string metricName;\n");
+        outs("\tmetricName.reserve(100);\n");
```
I don't think this reserve buys you much?  get_metric_name()  does multiple appends to a locally declared string and then returns it by value... the return might be optimized by the compiler?   But I think if you did the reserve and then passed the string by reference into get_metric_name() then it might be more efficient doing those multiple appends directly into this variable with the length reserved?

kenrowland replied:
Correct. Before ensuring all generated code was ifdef'd out, I had an empty method, so I needed to return something, hence the local variable. Now that no code is created if not instrumented, I changed it to pass a reference as indicated.

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -5757,26 +5792,59 @@ void EspServInfo::write_esp_binding()
+        outf("\nstd::string C%sSoapBinding::get_metric_name(const char *processName, const char *serviceName, const char *methodName)\n",
+             name_);
+        outs("{\n");
+        //outs("\tstatic std::string illegalChars = \"_\"\n;");
```
Can you remove the commented out lines?

kenrowland replied:
removed.

afishbeck reviewed (commented):
@kenrowland some new comments / possible change.

afishbeck reviewed (approved):
Looks much cleaner this way.

kenrowland commented:
@ghalliday or @richardkchapman Please merge

ghalliday reviewed (commented):
@kenrowland one minor issue

kenrowland commented:
@ghalliday This should be ready for merge.

ghalliday commented on esp/espcommon/espcommon.cpp:
```diff
@@ -52,8 +52,12 @@ static std::vector<__uint64> getExecutionProfileBuckets(const char *profileOptio
 
-// Registers a profiling metric for an ESP service method
-ESPCOMMON_API std::shared_ptr<hpccMetrics::ScaledHistogramMetric> registerProfilingMetric(const char *histogramMetricName, const char *desc, const char *profilingOptions)
+// Generate metric name and register it for the indicated service method
```
Note: You will probably require the original function later on.  This should probably be an additional function which calls the old one.

ghalliday commented on system/jlib/jmetrics.cpp:
```diff
@@ -138,35 +139,49 @@ bool MetricsManager::addMetric(const std::shared_ptr<IMetric> &pMetric)
-    {
-        name.append(".").append(metaDataIt.value);
-    }
+    std::regex pattern("^[A-Za-z][A-Za-z0-9.]*[A-Za-z0-9]$");
```
trivial: This should be a member of MetricManager, so that the pattern object can be created once (which can be expensive), and reused by all calls.  If this is going to be called for large numbers it may become significant.

ghalliday reviewed (commented):
@kenrowland a few trivial issues from a final check.  The regex moving within the class is may be significant enough to change.  I can merge as-is if you prefer.

kenrowland commented:
@ghalliday I went ahead and addressed your comments instead of waiting for another Jira. This code should be as clean as possible from the start. Please take look.

ghalliday reviewed (approved):
@kenrowland please squash

kenrowland commented:
@ghalliday sorry for delay, was on PTO. Ready to be merged.
